[CS-4721] Logic for restore wallet with seed phrase

oidoug opened the pull request:
### Description

This PR adds the option to import a wallet using a seed phrase. It uses parts of soon deprecated `useImportSeedSheet` and abstracts calls to derive wallet and ENS address to `ethers-wallet` model.

- [x] Completes #(CS-4721)

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

### Screenshots

Vid available on linear.

oidoug committed:
Adds model for wallet and ens derivation from seed;
Adds hook to abstract wallet import from seed.

oidoug committed:
Use of refs to keep wallet and ens checked info;
Adds navigationKey to backup stack so it updates when hasWallet changes value.

oidoug committed:
Adds loading state to import (done) buttom

oidoug committed:
Revert navigation to original

danibonilha commented on cardstack/src/hooks/backup/useWalletSeedPhraseImport.ts:
```diff
@@ -0,0 +1,76 @@
+        type: 'wallet_profile',
+        withoutStatusBar: true,
+      });
+    },
```
Does it make sense to keep using this import modal ?

oidoug replied:
Probably not. We moved away from these account customizations in favor of profiles.

danibonilha commented on cardstack/src/models/ethers-wallet.ts:
```diff
@@ -62,3 +67,56 @@ const getEthersWallet = async (accountAddress?: string) => {
+  } catch (error) {
+    logger.log('Error looking up ENS for imported HD type wallet', error);
+  }
+};
```
not sure we need this ens stuff anymore, if we are never showing the user name, I'd suggest to once we hit the Done button on the seed screen, we start importing and remove on this old profile related stuff

oidoug replied:
Ok, I agree. I'll not use the import sheet here then. 

Better also revert the changes I did on useImportSeedSheet, leave it as deprecated.

oidoug committed:
Removes Import wallet modal screen;
Reverts changes in useImportSeedSheet - will be deprecated;
Cleans unneeded code.

oidoug committed:
Cleans unneeded imports in ethers-wallet;
Reverts navigation.

oidoug committed:
Improves abstraction on validating seed phrase

oidoug committed:
Adds tests for useWalletSeedPhraseImport;
Adds tests for useBackupRestorePhraseScreen.

oidoug committed:
Adds clearAllMocks, otherwise other tests randomly fail

oidoug committed:
Mocks useWalletSeedPhraseImport use on tests for useBackupRestorePhraseScreen to avoid strange tests failing issue.

danibonilha commented on cardstack/src/hooks/backup/useWalletSeedPhraseImport.ts:
```diff
@@ -0,0 +1,47 @@
+) => {
+  const { importWallet } = useWalletManager();
+
+  const checkedWallet = useRef<EthereumWalletFromSeed | undefined>(ethWallet);
```
checkedWallet is only used inside `handleImportWallet`. So there's no need to create a ref here, we can scope it inside the function.

oidoug replied:
My intention was to reuse this hook in `ImportSeedSheet` but went another way after a previous comment from you. I'll simplify this.

danibonilha commented on cardstack/src/hooks/backup/useWalletSeedPhraseImport.ts:
```diff
@@ -0,0 +1,47 @@
+
+  const sanitizedSeedPhrase = useMemo(() => sanitizeSeedPhrase(seedPhrase), [
+    seedPhrase,
+  ]);
```
same for this const, only used inside `handleImportWallet`

danibonilha commented on cardstack/src/hooks/backup/useWalletSeedPhraseImport.ts:
```diff
@@ -0,0 +1,47 @@
+  ]);
+
+  const handleImportWallet = useCallback(async () => {
+    if (!isSeedPhraseValid || !sanitizedSeedPhrase) {
```
what does a `not sanitizedSeedPhrase` mean ? Hm, maybe we should add this check within the `isSeedPhraseValid` . Oh I think `isValidSeedPhrase` already sanitizes it, so not sure we need to do this here

oidoug replied:
You're right, it's redundant, I'll clean this up.

danibonilha commented on cardstack/src/models/ethers-wallet.ts:
```diff
@@ -62,3 +65,26 @@ const getEthersWallet = async (accountAddress?: string) => {
+): Promise<EthereumWalletFromSeed | undefined> => {
+  if (!seedPhrase || !isValidSeedPhrase(seedPhrase)) return undefined;
+
+  const cleanSeedPhrase = sanitizeSeedPhrase(seedPhrase);
```
we are sanitizing it before calling this funciton, and then we sanitize again here ?

oidoug replied:
This was supposed to be reused, doesn't add value to what we already have in `ethereumUtils` export. I'll remove this extra function out 👍

danibonilha commented on cardstack/src/models/ethers-wallet.ts:
```diff
@@ -62,3 +65,26 @@ const getEthersWallet = async (accountAddress?: string) => {
+  seedPhrase: string
+): Promise<EthereumWalletFromSeed | undefined> => {
+  if (!seedPhrase || !isValidSeedPhrase(seedPhrase)) return undefined;
+
```
I'm confused about this function, there's this check on the handle import, and then we are doing the same checks here again, also this is just used while importing the wallet

oidoug replied:
ditto

danibonilha commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,59 @@
+  const onDonePressed = useCallback(async () => {
+    setLoading();
+
+    if (isSeedPhraseValid) {
+      await handleImportWallet();
+    }
```
if we check when seed is valid here, maybe we don't need to check on `handleImportWallet`

danibonilha commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,59 @@
+    (updatedPhrase: string) => {
+      setPhrase(updatedPhrase);
+
+      const words = updatedPhrase.split(' ');
+
+      // Phrase is deamed complete when it has 12 words
+      const isComplete =
+        words.length === 12 && words[words.length - 1].length > 0;
+
```
This is partially the same thing that is done on `isValidSeedPhrase`  so maybe we could add another function that partially validates the format

danibonilha replied:
Or you could use this regex `updatedPhrase.match(/\w+/g).length` then, there's no need to trim, because \w already gets all words and ignores the spaces and punctuations.

Commit:
Merge branch 'develop' into doug/cs-4721-logic-restore-wallet-with-seed-phrase

danibonilha commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,59 @@
+  );
+
+  const onDonePressed = useCallback(async () => {
+    setLoading();
```
What is this loading for ? Doesn't importWallet already shows the loading overlay ? 🤔

oidoug replied:
It was useful before when I was checking for ens - it took a while an a loading state was needed for the done button. It's not needed anymore, I'll remove it.

danibonilha commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,59 @@
+
+  const { handleImportWallet, isSeedPhraseValid } = useWalletSeedPhraseImport(
+    phrase
+  );
```
I'm even wondering if we need this to be a separated hook, technically the importWallet will only happen on this screen, so maybe we could move it's logic to here, since it will basically be this single `handleImportWallet` function

oidoug replied:
Yeah kinda overthinking on my end.
Not sure how we'll handle importing from settings, it will still be there for now I think, but in any case, it's not using this extra hook now anyways.

danibonilha replied:
I guess we don't need to import from settings anymore, it was just something useful for devs while we still had multi-wallet support, so i'm good with removing it if y'all agree to.

paulinhapenedo commented on cardstack/src/hooks/backup/useWalletSeedPhraseImport.ts:
```diff
@@ -0,0 +1,47 @@
+import { sanitizeSeedPhrase } from '@rainbow-me/utils/formatters';
+import logger from 'logger';
+
+export const useWalletSeedPhraseImport = (
+  seedPhrase: string,
+  ethWallet?: EthereumWalletFromSeed
+) => {
+  const { importWallet } = useWalletManager();
+
+  const checkedWallet = useRef<EthereumWalletFromSeed | undefined>(ethWallet);
```
Why are we receiving a wallet object as param? I don't see us using it anywhere in the code right now. Am I missing something?

Removing this param could simplify the code by removing the `useRef` and using a regular `const` in line 34.

oidoug replied:
My intention was to reuse this hook in `ImportSeedSheet` but went another way after a comment from Dani. I'll simplify this.

oidoug committed:
Simplifies hooks for restore with phrase;
Removes unecessary loading state;

oidoug committed:
Fixes import of sanitizeSeedPhrase for test

oidoug committed:
Refactors wallet import into useBackupRestorePhraseScreen hook and updates testsl
Deletes now unecessary useWalletSeedPhraseImport hook;

paulinhapenedo commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,77 @@
+      return;
+    }
+
+    console.log(':::1');
```
Ooops ✂️

paulinhapenedo commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,77 @@
+    try {
+      const sanitizedPhrase = sanitizeSeedPhrase(phrase);
+
+      console.log(':::2');
```
Ooops ✂️

paulinhapenedo commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,77 @@
+        sanitizedPhrase
+      );
+
+      console.log(':::3', { sanitizedPhrase, checkedWallet });
```
Ooops ✂️

oidoug replied:
🤦  my bad

oidoug committed:
Removing leftover comments

oidoug committed:
Fixes Phrase Input keyboard not bluring on submit.

paulinhapenedo commented on cardstack/src/screens/Backup/BackupRestorePhraseScreen/useBackupRestorePhraseScreen.ts:
```diff
@@ -1,13 +1,71 @@
+        seed: sanitizedPhrase,
+        checkedWallet,
+      });
+    } catch (error) {
+      logger.error('Error deriving and importing wallet.', error);
```
nit: shouldn't we be logging this into Sentry?

oidoug replied:
I wondered the same but was afraid what could eventually be returned in `error` and turn this into a privacy issue.

paulinhapenedo reviewed (approved):
Looks good! Just a nit comment that you can feel free to ignore.
